### PR TITLE
Make API friendly to block explorers

### LINKED ISSUE #694: API additions required for blockchain explorers
I had a look on your REST API in order to evaluate the feasibility to plug lighthouse into our open source block explorer beaconcha.in. Most of the API already provides most of the information required for the explorer but some crucial details are still missing:

- The blocks object should also include the block root hash alongside the state root hash and parent root hash
- Currently I could not find any call that returns the current balance of a validator, it would be great if it is included in the get_beacon_validators response alongside the effective balance
- The get_beacon_validators should include the validator index alongside the validator public key
- Add a get_beacon_validators_duties that returns the duties of all validators for a given epoch, otherwise we will have to call get_beacon_validator_duties with all public keys which might be not that performant (but I haven't tested this up to now)
- In order to de-aggregate attestations and associate an attestation with the included validators we will need the beacon committees for each epoch. I couldn't find any call that returns the committees by epoch
- In order to show participation statistics for each epoch add a call that shows the eligible ether, voted ether and the participation rate for each epoch

With those few changes we should be able to write a thin compatibility wrapper and have a dedicated instance of the lighthouse testnet available at https://beaconcha.in/

For optimal compatibility see the protobuf definitions provided by prysmaticlabs at https://github.com/prysmaticlabs/ethereumapis/

### LINKED ISSUE #713: JSON for public key does not have 0x prefix
## Description

The `0x` prefix is not being added when serializing a pubkey as JSON. As an example:

```
"pubkey": "98f87bc7c8fa10408425bbeeeb3dc387e3e0b4bd92f57775b60b39156a16f9ec80b273a64269332d97bdb7d93ae05a16"
```

## Steps to resolve

Ensure that JSON encoding and decoding expects a `0x` prefix for public keys.

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

- Resolves #694
- Resolves #713

## Proposed Changes

Point-by-point addressing of @ppratscher's requests:

> The blocks object should also include the block root hash alongside the state root hash and parent root hash

This has been done, see [/beacon/block](https://github.com/sigp/lighthouse/blob/3b084db59507de3c7467fb18c104da1215ac7064/book/src/http_beacon.md#beaconblock).

> Currently I could not find any call that returns the current balance of a validator, it would be great if it is included in the get_beacon_validators response alongside the effective balance

The [`/beacon/validators`](https://github.com/sigp/lighthouse/blob/3b084db59507de3c7467fb18c104da1215ac7064/book/src/http_beacon.md#beaconvalidators) endpoint now includes the balance.

> The get_beacon_validators should include the validator index alongside the validator public key

The [`/beacon/validators`](https://github.com/sigp/lighthouse/blob/3b084db59507de3c7467fb18c104da1215ac7064/book/src/http_beacon.md#beaconvalidators) endpoint now includes the validator index.

> Add a get_beacon_validators_duties that returns the duties of all validators for a given epoch, otherwise we will have to call get_beacon_validator_duties with all public keys which might be not that performant (but I haven't tested this up to now)

The [`/beacon/validators/all`](https://github.com/sigp/lighthouse/blob/3b084db59507de3c7467fb18c104da1215ac7064/book/src/http_beacon.md#beaconvalidatorsall) endpoint serves this request. It's returning quite quickly on my end (40ms).

> In order to de-aggregate attestations and associate an attestation with the included validators we will need the beacon committees for each epoch. I couldn't find any call that returns the committees by epoch

The [`/beacon/committees`](https://github.com/sigp/lighthouse/blob/bitfly/book/src/http_beacon.md#beaconcommittees) endpoint should fulfill this request.

> In order to show participation statistics for each epoch add a call that shows the eligible ether, voted ether and the participation rate for each epoch

The [`consensus/global_votes`](https://github.com/sigp/lighthouse/blob/3b084db59507de3c7467fb18c104da1215ac7064/book/src/http_consensus.md#consensusglobal_votes) endpoint should serve this request.

- eligible ether = `previous_epoch_active_gwei`
- voted ether = `previous_epoch_attesting_gwei`
- participation rate = voted ether / eligible ether (I didn't see the need to include this in the endpoint)

## Notes

~~This is a work-in-progress.~~

### CODE REVIEW

**Comment by paulhauner** [comment]
@ppratscher, I have finished implementing the endpoints you were after and I have addressed each point in the top of this PR.

I will move onto unit testing next, however I have tested the endpoints by hand. If you are eager, feel free to start building off this branch. Otherwise, feel free to wait until it has been merged to master.

**Comment by ppratscher** [comment]
@paulhauner thanks a lot for your efforts! we will be starting next week on Monday to add a compatibility layer to beaconcha.in in order to support lighthouse. Will report back in case we run into any issues.

**Comment by pscott** [comment]
Just read the docs in this PR, everything looked good to me! Disappointed I couldn't find a single typo... :p

File: book/src/http_consensus.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+## `/consensus/global_votes`
+
+Returns a global count of votes for some given `epoch`. The results are included
+both for the current and previous (`epoch - 1`) epochs since both are required
+by the beacon node whilst performing per-epoch-processing.
+
+Generally, you should consider the "current" values to be incomplete and the
+"previous" values to be final. This is because validators can continue to
+include attestations from the _current_ epoch in the _next_ epoch, however this
+is not the case for attestations from the _previous_ epoch.
```
Comment (michaelsproul) [inline book/src/http_consensus.md]: I was confused by this paragraph until I looked at the code and saw that the `epoch` being requested must have already come to an end, i.e. you can't request data midway through the current wall-clock epoch. Perhaps this API would be better returning just one set of balances, which it derives from the `previous_` fields of the following epoch's `ValidatorStatuses` struct. So if the current wall-clock epoch is `N`, you can use this function to request finalized voting information for epochs up to `N - 2`. That has the disadvantage of not being able to see what happened in the previous wall-clock epoch... but with the current design we already can't see what's happening in the current wall-clock epoch. Perhaps there could be two API endpoints? One for "finalized" data from all epochs up to `N - 2` and another for "live" info from the current and previous epochs?

File: book/src/http_network.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+### Example Response
+
+```json
+"QmVFcULBYZecPdCKgGmpEYDqJLqvMecfhJadVBtB371Avd"
+```
+
+## `/network/peers`
+
+Requests the beacon node for one `MultiAddr` for each connected peer.
```
Comment (michaelsproul) [inline book/src/http_network.md]: Gramatically awkward, maybe:

> Requests one `MultiAddr` for each peer connected to the beacon node

File: book/src/http_network.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+```json
+[
+	"QmaPGeXcfKFMU13d8VgbnnpeTxcvoFoD9bUpnRGMUJ1L9w",
+	"QmZt47cP8V96MgiS35WzHKpPbKVBMqr1eoBNTLhQPqpP3m"
+]
+```
+
+## `network/enr`
+
+Requests the beacon node for it's listening `ENR` address.
```
Comment (michaelsproul) [inline book/src/http_network.md]: its

File: book/src/http_consensus.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+"previous_epoch" and "target_attester".
+
+### HTTP Specification
+
+| Property | Specification |
+| --- |--- |
+Path | `/consensus/individual_votes`
+Method | GET
+JSON Encoding | Object
+Query Parameters | `epoch`
```
Comment (pawanjay176) [inline book/src/http_consensus.md]: This should be a `POST` method with a request body.

File: book/src/http_consensus.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+Typical Responses | 200
+
+### Parameters
+
+Requires the `epoch` (`Epoch`) query parameter to determine which epoch will be
+considered the current epoch.
```
Comment (pawanjay176) [inline book/src/http_consensus.md]: I think this should be a `RequestBody` with `epoch` and `pubkeys` as parameters.

File: book/src/http_consensus.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+                              |     |
+							  v     v
+Epoch:  |---previous---|---current---|---next---|
+
+                          |-------------|
+						         ^
+                                 |
+		       window for including "current" attestations
+					        in a block
+```
```
[the remark this answers is not in the document]
Comment (pscott) [reply book/src/http_consensus.md]: This was something I was wondering too but from what I understood it's actually ok? The window spans on `current` and a little bit on `next` right?

**Comment by ppratscher** [comment]
Thanks a lot for your efforts. Today we completed the integration of the lighthouse rpc api into our block explorer. Must of the calls work as expected, we just found 1 oddity:

For some states calling `/beacon/validators/all` returns `Unable to read pubkey cache: PubkeyCacheIncomplete { cache_len: 0, registry_len: 16384 }`

An example would be retrieving the validators with the state root of slot 192:
`http://localhost:5052/beacon/validators/all?state_root=0xdf0b538dbf80d2a1d1933b80052a740e1a4f07297f83c8948650977ecf09d418`

File: book/src/http_consensus.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (paulhauner) [reply book/src/http_consensus.md]: Unfortunately it renders well in either github **or** the book and we have to choose one. I chose the book.

File: book/src/http_consensus.md (hunk 1 above)
Comment (paulhauner) [reply book/src/http_consensus.md]: As discussed with @michaelsproul, we'll leave this as-is for now.

**Comment by paulhauner** [comment]
> For some states calling /beacon/validators/all returns Unable to read pubkey cache: PubkeyCacheIncomplete { cache_len: 0, registry_len: 16384 }

Thanks @ppratscher, I know exactly what this is and will fix it.

**Comment by paulhauner** [comment]
All comments have been addressed :)

**Pull request merged**